util/resolver/resolver.go

```diff
@@ -192,3 +187,33 @@ func hostsWithCredentials(ctx context.Context, hosts docker.RegistryHosts, sm *s
 		return res, nil
 	}
 }
+
+func newDefaultClient() *http.Client {
+	return &http.Client{
+		Transport: newDefaultTransport(),
+	}
+}
+
+// newDefaultTransport is for pull or push client
+//
+// NOTE: For push, there must disable http2 for https because the flow control
+// will limit data transfer. The net/http package doesn't provide http2 tunable
+// settings which limits push performance.
+//
+// REF: https://github.com/golang/go/issues/14077
+func newDefaultTransport() *http.Transport {
+	return &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		MaxIdleConns:          10,
+		IdleConnTimeout:       30 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 5 * time.Second,
+		DisableKeepAlives:     true,
+		TLSNextProto:          make(map[string]func(authority string, c *tls.Conn) http.RoundTripper),
```

Comment:
So is it `TLSNextProto` or `DisableKeepAlives` that is important in here? In Docker I don't see call to `TLSNextProto` so not sure why that isn't affected then.

Reply:
The `onceSetNextProtoDefaults` will be called once at the beginning of `RoundTrip`, which used to setup http2 RoundTrip transport for https request if possible. And the `TLSNextProto` is mentioned in `net/http` package document. But there are several way to disable http2, for instance, dockerd uses non-nil `TLSClientConfig` to avoid http2.

> roundTrip in [golang1.13](https://github.com/golang/go/blob/release-branch.go1.13/src/net/http/transport.go#L454-L456) and [golang1.14](https://github.com/golang/go/blob/release-branch.go1.14/src/net/http/transport.go#L484-L486)
> onceSetNextProtoDefaults in [golang1.13](https://github.com/golang/go/blob/release-branch.go1.13/src/net/http/transport.go#L348-L361) and [golang1.14](https://github.com/golang/go/blob/release-branch.go1.14/src/net/http/transport.go#L364-L377)
```
// roundTrip implements a RoundTripper over HTTP.
func (t *Transport) roundTrip(req *Request) (*Response, error) {
	t.nextProtoOnce.Do(t.onceSetNextProtoDefaults)
        ...
}

func (t *Transport) onceSetNextProtoDefaults() {
...
        if t.TLSNextProto != nil {
		// This is the documented way to disable http2 on a
		// Transport.
		return
	}
	if !t.ForceAttemptHTTP2 && (t.TLSClientConfig != nil || t.Dial != nil || t.DialContext != nil || t.hasCustomTLSDialer()) {
		// Be conservative and don't automatically enable
		// http2 if they've specified a custom TLS config or
		// custom dialers. Let them opt-in themselves via
		// http2.ConfigureTransport so we don't surprise them
		// by modifying their tls.Config. Issue 14275.
		// However, if ForceAttemptHTTP2 is true, it overrides the above checks.
		return
	}
....
}
```

If the http2 is enabled, the http2 RoundTripp transport will use `http2clientStream` to write request body to server.

> writeRequestBody in [golang1.13](https://github.com/golang/go/blob/release-branch.go1.13/src/net/http/h2_bundle.go#L7710) and [golang1.14](https://github.com/golang/go/blob/release-branch.go1.14/src/net/http/h2_bundle.go#L7736)
```
func (cs *http2clientStream) writeRequestBody(body io.Reader, bodyCloser io.Closer) (err error) {
	cc := cs.cc
	sentEnd := false // whether we sent the final DATA frame w/ END_STREAM
	buf := cc.frameScratchBuffer()
	defer cc.putFrameScratchBuffer(buf)

	....

	var sawEOF bool
	for !sawEOF {
		n, err := body.Read(buf)
		if err == io.EOF {
			sawEOF = true
			err = nil
		} else if err != nil {
			cc.writeStreamReset(cs.ID, http2ErrCodeCancel, err)
			return err
		}

		remain := buf[:n]
		for len(remain) > 0 && err == nil {
			var allowed int32
			allowed, err = cs.awaitFlowControl(len(remain))
			...
			cc.wmu.Lock()
			data := remain[:allowed]
			remain = remain[allowed:]
			sentEnd = sawEOF && len(remain) == 0 && !hasTrailers
			err = cc.fr.WriteData(cs.ID, sentEnd, data)
			...
			cc.wmu.Unlock()
		}
		
	}
        ...
}

```

For writing data frame, the client needs to wait for available buffer from flow control by `awaitFlowControl`. The buffer from flow control will be updated by `WINDOW_UPDATE` frame which will be send by server when it receives data frame. However, the increase for buffer is too small and slow than client uploads. Sometimes, the client needs to wait for 200ms to write 30kb data. It is unacceptable.

For http1, just use `io.Copy` to upload data and let TCP proto to handle flow control. 

> writeBody in [golang1.13](https://github.com/golang/go/blob/release-branch.go1.13/src/net/http/transfer.go#L359) and [golang1.14](https://github.com/golang/go/blob/release-branch.go1.14/src/net/http/transfer.go#L359)

```
func (t *transferWriter) writeBody(w io.Writer) error {
	var err error
	var ncopy int64

	// Write body. We "unwrap" the body first if it was wrapped in a
	// nopCloser. This is to ensure that we can take advantage of
	// OS-level optimizations in the event that the body is an
	// *os.File.
	if t.Body != nil {
		var body = t.unwrapBody()
		...
		} else {
			ncopy, err = t.doBodyCopy(w, io.LimitReader(body, t.ContentLength))
			if err != nil {
				return err
			}
			var nextra int64
			nextra, err = t.doBodyCopy(ioutil.Discard, body)
			ncopy += nextra
		}
		...
	}
```   

cc @dmcgowan

Reply:
The golang http2 just setup 64KB buffer for flow control and client needs to wait for WINDOW_UPDATE frame to update the buffer before upload. The init buffer size cannot be configurable. it is painful.

And the docker/distribution is setting TLS on listener not for http.Server so that the server will setup http2 by default for https request.